sync/src/synchronizer/block_fetcher.rs

```diff
@@ -151,9 +141,20 @@ impl BlockFetcher {
             let mut header = self
                 .active_chain
                 .get_ancestor(&best_known_header.hash(), start + span - 1)?;
+
+            // Judge whether we should fetch the target block, neither stored nor in-flighted
             for _ in 0..span {
                 let parent_hash = header.parent_hash();
-                if should_fetch(header.hash()) {
+                // NOTE: Filtering `BLOCK_STORED` but not `BLOCK_RECEIVED`, is for avoiding
+                // stopping synchronization even when orphan_pool maintains dirty items by bugs.
+                let stored = self
+                    .active_chain
+                    .contains_block_status(&parent_hash, BlockStatus::BLOCK_STORED);
+                if stored {
+                    // If the block is stored, its ancestor must on store
+                    // So we can skip the search of this space directly
+                    break;
```

Comment:
Should we break outer while as well?

Reply:
> Should we break outer while as well?

No, this logic is segmented, the parent here is just the current segment can be ignored. 

something like:

```
|  first | second | ....
|  <--   |   <--  |....
```